experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/editableTree.events.spec.ts

```diff
@@ -0,0 +1,413 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { strict as assert } from "assert";
+
+import { MockFluidDataStoreRuntime } from "@fluidframework/test-runtime-utils";
+
+import { FieldKinds, SchemaBuilder } from "../../../feature-libraries";
+import { AllowedUpdateType } from "../../../core";
+import { TypedTreeFactory } from "../../../typed-tree";
+import { ForestType } from "../../../shared-tree";
+import { typeboxValidator } from "../../../external-utilities";
+import { leaf } from "../../..";
+
+describe("beforeChange/afterChange events", () => {
+	const builder = new SchemaBuilder({
+		scope: "beforeChange/afterChange events",
+		libraries: [leaf.library],
+	});
+	const myInnerNodeSchema = builder.struct("myInnerNode", {
+		myInnerString: SchemaBuilder.required(leaf.string),
+	});
+	const myNodeSchema = builder.structRecursive("myNode", {
+		child: SchemaBuilder.required(myInnerNodeSchema),
+		myString: SchemaBuilder.required(leaf.string),
+		myOptionalNumber: SchemaBuilder.optional(leaf.number),
+	});
+	const schema = builder.toDocumentSchema(SchemaBuilder.field(FieldKinds.required, myNodeSchema));
+
+	it("fire the expected number of times", () => {
+		// TODO: once assignment to properties is implemented in EditableTree2, update this test to apply changes like
+		//   root.myString = "new string";
+		// instead of
+		//   root.boxedMyString.content = "new string";
+
+		const factory = new TypedTreeFactory({
+			jsonValidator: typeboxValidator,
+			forest: ForestType.Reference,
+			allowedSchemaModifications: AllowedUpdateType.None,
+			initialTree: {
+				myString: "initial string",
+				myOptionalNumber: 3,
+				child: { myInnerString: "initial string in child" },
+			},
+			schema,
+			subtype: "test",
+		});
+
+		const root = factory.create(new MockFluidDataStoreRuntime(), "the tree").root.content;
+
+		let rootBeforeChangeCount = 0;
+		let rootAfterChangeCount = 0;
+		let childBeforeChangeCount = 0;
+		let childAfterChangeCount = 0;
+
+		root.on("beforeChange", (upPath) => {
+			rootBeforeChangeCount++;
+		});
+		root.on("afterChange", (upPath) => {
+			rootAfterChangeCount++;
+		});
+
+		assert.strictEqual(rootBeforeChangeCount, 0);
+		assert.strictEqual(rootAfterChangeCount, 0);
+
+		// Replace existing node - myString; should fire events on the root node.
+		root.boxedMyString.content = "new string";
+
+		assert.strictEqual(rootBeforeChangeCount, 1);
+		assert.strictEqual(rootAfterChangeCount, 1);
+
+		// Add node where there was none before - child; should fire events on the root node.
```

Comment:
Why does this say "where there was none before"? Isn't that part of the tree initialized with `{ myInnerString: "initial string in child" }`?

Reply:
Thanks for catching that, this test should indeed initialize the tree with an empty `child`. Fixed locally, I'll push it soon.